File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if not s or not words:
            return []

        word_length = len(words[0])
        num_words = len(words)
        total_length = word_length * num_words
        word_count = Counter(words)  
        result = []

        for i in range(word_length):
            left, right = i, i
            curr_count = Counter()
            while right + word_length <= len(s):
                word = s[right:right + word_length]  # Ambil kata dari substring
                right += word_length

                if word in word_count:
                    curr_count[word] += 1

                    while curr_count[word] > word_count[word]:
                        curr_count[s[left:left + word_length]] -= 1
                        left += word_length

                    if right - left == total_length:
                        result.append(left)
                else:
                    curr_count.clear()
                    left = right  # Reset window

        return result

File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.java
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();

        int wordLen = words[0].length();
        int permutationLen = wordLen * words.length;

        Map<String, Integer> wordCount = new HashMap<>();
        Map<String, Integer> windowCount = new HashMap<>();

        for (String word : words){
            wordCount.put(word, wordCount.getOrDefault(word,0) + 1);
        }

        // Offset e.g wordLen == 2 => 0,2,4,... & 1,3,5,...
        for (int i = 0; i < wordLen; i++){
            int left = i;
            int right = i;
            int count = 0;
            
            while (right + wordLen <= s.length()){
                String word = s.substring(right, right + wordLen);
                
                if (wordCount.containsKey(word)){
                    windowCount.put(word, windowCount.getOrDefault(word,0) + 1);
                    count += wordLen;

                    while (windowCount.get(word) > wordCount.get(word)){
                        String leftWord = s.substring(left, left + wordLen);
                        windowCount.put(leftWord, windowCount.get(leftWord) - 1);
                        left += wordLen;
                        count -= wordLen;
                    }

                    if (count == permutationLen){
                        result.add(left);
                    }
                    
                } else {
                    windowCount.clear();
                    count = 0;
                    left = right + wordLen;
                }

                System.out.println(word);
                System.out.println(count);
                System.out.println();
                right += wordLen;
            }
        }

        

        return result;
    }
}